In [ ]:
import pandas as pd
original_dataset=pd.read_json('https://raw.githubusercontent.com/usnistgov/SelfishMiningSim/master/all_results.json')
original_dataset.head(5)

daa                              difficulty_parameters  \
0   btc                 {'period': 2016, 'offbyone': True}   
1  dash              {'npastblocks': 24, 'offbyone': True}   
2   bch  {'lookback': 144, 'offbyone': True, 'mediantim...   
3   btc                {'period': 2016, 'offbyone': False}   
4   zec  {'navginterval': 17, 'nmediantimespan': 11, 'n...   

                                             results  
0  [{'numsims': 30, 'alpha': 0.2, 'gamma': 0, 'ti...  
1  [{'numsims': 30, 'alpha': 0.06, 'gamma': 0, 't...  
2  [{'numsims': 30, 'alpha': 0.06, 'gamma': 0, 't...  
3  [{'numsims': 30, 'alpha': 0.2, 'gamma': 0, 'ti...  
4  [{'numsims': 30, 'alpha': 0.06, 'gamma': 0, 't...

In [ ]:
df=original_dataset
diffparam=df['difficulty_parameters'].apply(pd.Series)
list_df=[]
for i in range(len(df)):
    df1=df['results'].iloc[i]
    data=pd.DataFrame(df1)
    data.insert(0,'daa',df['daa'].iloc[i])
    diff=df['difficulty_parameters'].iloc[i]
    row=pd.DataFrame(diffparam.iloc[i].dropna())
    row=row.transpose()
    for pos,key in enumerate(diff.keys()):
        data.insert(pos,key,row[key].iloc[0])
    list_df.append(data)
len(list_df)

#list_df variable containts every row of the original dataset expanded as a single dataset.
#To access each dataset, just iterate through the list elements.

149

In [ ]:
for temp in list_df:
    print(temp.shape)

(855, 22)
(1254, 22)
(1254, 23)
(855, 22)
(1254, 25)
(1254, 22)
(1254, 23)
(855, 22)
(1254, 22)
(1254, 22)
(1254, 22)
(1254, 22)
(1254, 22)
(1254, 22)
(1254, 22)
(1254, 22)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 23)
(1254, 22)
(1254, 22)
(78, 23)
(660, 25)
(660, 25)
(660, 25)
(660, 25)
(660, 25)
(660, 25)
(660, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)
(264, 25)

The Above program was coded to extract 149 different dataframes based on every row as discussed .
But upon my analysis of the dataset, every algorithm has a unique set of difficulty parameters. So i think it would be better if we come up with 6 datasets(One for each algorithm)
But in that case the number of rows in each dataset will be higher.
I have Coded the same below, Just Check once if this will do

In [ ]:
def extract(algo):
    df=original_dataset[original_dataset['daa']==algo]
    diffparam=df['difficulty_parameters'].apply(pd.Series)
    df1=df['results'].iloc[0]
    dataframe=pd.DataFrame(df1)
    diff=df['difficulty_parameters'].iloc[0]
    row=pd.DataFrame(diffparam.iloc[0].dropna())
    row=row.transpose()
    for pos,key in enumerate(diff.keys()):
        dataframe.insert(pos,key,row[key].iloc[0])
    for i in range(1,len(df)):
        df1=df['results'].iloc[i]
        data=pd.DataFrame(df1)
        diff=df['difficulty_parameters'].iloc[i]
        row=pd.DataFrame(diffparam.iloc[i].dropna())
        row=row.transpose()
        for pos,key in enumerate(diff.keys()):
            data.insert(pos,key,row[key].iloc[0])
        dataframe=pd.concat([dataframe,data],ignore_index=True)
    dataframe.insert(0,'daa',df['daa'].iloc[0])
    return dataframe

In [ ]:
algorithms=list(original_dataset['daa'].unique())
dataset=dict()
for algo in algorithms:
    dataset[algo]=extract(algo)
#The Dataset Variable is a dictonary containing dataset extracted for each algorithm with the key as the algorithm's name i.e daa
#Given below is a demo of what each dataset will contain.  

(9234, 22)

In [ ]:
print(dataset['btc'].shape)
dataset['btc'].head()

(9234, 22)


daa  period  offbyone  numsims  alpha  gamma  timewarp  numblocks  \
0  btc    2016      True       30    0.2   0.00         0      10000   
1  btc    2016      True       30    0.2   0.00      3600      10000   
2  btc    2016      True       30    0.2   0.00      7200      10000   
3  btc    2016      True       30    0.2   0.05         0      10000   
4  btc    2016      True       30    0.2   0.05      3600      10000   

   blocktime  winratio  ...  relativegain  adjustedrelativegain  gainstddev  \
0        600  0.093275  ...     -0.533623             -0.551429         0.0   
1        600  0.092009  ...     -0.539955             -0.558573         0.0   
2        600  0.093352  ...     -0.533241             -0.552124         0.0   
3        600  0.101965  ...     -0.490173             -0.510206         0.0   
4        600  0.101375  ...     -0.493124             -0.513377         0.0   

   adjustedgainsteddev  secondsperblockstddev   finalheight    numreorgs  \
0                  0.0                    0.0  11000.200000  1531.433333   
1                  0.0                    0.0  11000.566667  1515.866667   
2                  0.0                    0.0  11000.233333  1514.266667   
3                  0.0                    0.0  11000.100000  1518.933333   
4                  0.0                    0.0  11000.133333  1517.300000   

   smwinreorgs  didbetternaive  didbettertimeadjust  
0     0.201882             0.0                  0.0  
1     0.198572             0.0                  0.0  
2     0.196568             0.0                  0.0  
3     0.201504             0.0                  0.0  
4     0.199095             0.0                  0.0  

[5 rows x 22 columns]

In [ ]:
print(dataset['dash'].shape)
dataset['dash'].head()

(7524, 22)


daa  npastblocks  offbyone  numsims  alpha  gamma  timewarp  numblocks  \
0  dash           24      True       30   0.06   0.00         0      10000   
1  dash           24      True       30   0.06   0.00      3600      10000   
2  dash           24      True       30   0.06   0.00      7200      10000   
3  dash           24      True       30   0.06   0.05         0      10000   
4  dash           24      True       30   0.06   0.05      3600      10000   

   blocktime  winratio  ...  relativegain  adjustedrelativegain  gainstddev  \
0        150  0.013402  ...     -0.776634             -0.787508           0   
1        150  0.024531  ...     -0.591154             -0.608640           0   
2        150  0.034366  ...     -0.427226             -0.432776           0   
3        150  0.015778  ...     -0.737026             -0.749869           0   
4        150  0.027010  ...     -0.549826             -0.568709           0   

   adjustedgainsteddev  secondsperblockstddev   finalheight  numreorgs  \
0                    0                      0  11000.033333      536.3   
1                    0                      0  11000.033333      505.4   
2                    0                      0  11000.033333      472.3   
3                    0                      0  11000.000000      531.1   
4                    0                      0  11000.066667      503.7   

   smwinreorgs  didbetternaive  didbettertimeadjust  
0     0.059112             0.0                  0.0  
1     0.108957             0.0                  0.0  
2     0.156436             0.0                  0.0  
3     0.059508             0.0                  0.0  
4     0.109000             0.0                  0.0  

[5 rows x 22 columns]

In [ ]:
print(dataset['zec'].shape)
dataset['zec'].head()

(31218, 25)


daa  navginterval  nmediantimespan  nmaxadjustup  nmaxadjustdown  \
0  zec            17               11            16              32   
1  zec            17               11            16              32   
2  zec            17               11            16              32   
3  zec            17               11            16              32   
4  zec            17               11            16              32   

   NPOWDampeningFactor  numsims  alpha  gamma  timewarp  ...  relativegain  \
0                    4       30   0.06   0.00         0  ...     -0.826074   
1                    4       30   0.06   0.00      3600  ...     -0.816741   
2                    4       30   0.06   0.00      7200  ...     -0.827851   
3                    4       30   0.06   0.05         0  ...     -0.782689   
4                    4       30   0.06   0.05      3600  ...     -0.783077   

   adjustedrelativegain  gainstddev  adjustedgainsteddev  \
0             -0.827025         0.0                  0.0   
1             -0.817775         0.0                  0.0   
2             -0.828805         0.0                  0.0   
3             -0.783908         0.0                  0.0   
4             -0.784338         0.0                  0.0   

   secondsperblockstddev   finalheight   numreorgs  smwinreorgs  \
0                    0.0  11000.100000  551.400000     0.060100   
1                    0.0  11000.033333  547.366667     0.061391   
2                    0.0  11000.000000  543.266667     0.059442   
3                    0.0  11000.033333  547.666667     0.060006   
4                    0.0  11000.000000  549.133333     0.060178   

   didbetternaive  didbettertimeadjust  
0             0.0                  0.0  
1             0.0                  0.0  
2             0.0                  0.0  
3             0.0                  0.0  
4             0.0                  0.0  

[5 rows x 25 columns]

In [ ]:
print(dataset['bch'].shape)
dataset['bch'].head()

(26334, 23)


daa  lookback  offbyone  mediantimepast  numsims  alpha  gamma  timewarp  \
0  bch       144      True               3       30   0.06   0.00         0   
1  bch       144      True               3       30   0.06   0.00      3600   
2  bch       144      True               3       30   0.06   0.00      7200   
3  bch       144      True               3       30   0.06   0.05         0   
4  bch       144      True               3       30   0.06   0.05      3600   

   numblocks  blocktime  ...  relativegain  adjustedrelativegain  gainstddev  \
0      10000        600  ...     -0.816575             -0.818561           0   
1      10000        600  ...     -0.818741             -0.820700           0   
2      10000        600  ...     -0.815130             -0.817145           0   
3      10000        600  ...     -0.777690             -0.780100           0   
4      10000        600  ...     -0.767469             -0.770003           0   

   adjustedgainsteddev  secondsperblockstddev   finalheight   numreorgs  \
0                    0                      0  11000.066667  551.833333   
1                    0                      0  11000.033333  548.333333   
2                    0                      0  11000.000000  541.233333   
3                    0                      0  11000.033333  546.266667   
4                    0                      0  11000.066667  545.266667   

   smwinreorgs  didbetternaive  didbettertimeadjust  
0     0.062662             0.0                  0.0  
1     0.061274             0.0                  0.0  
2     0.061658             0.0                  0.0  
3     0.059659             0.0                  0.0  
4     0.060982             0.0                  0.0  

[5 rows x 23 columns]

In [ ]:
print(dataset['xmr'].shape)
dataset['xmr'].head()

(10110, 23)


daa  lookback  delay  outliers  numsims  alpha  gamma  timewarp  numblocks  \
0  xmr       720     15        60       30   0.06   0.00         0      10000   
1  xmr       720     15        60       30   0.06   0.00      3600      10000   
2  xmr       720     15        60       30   0.06   0.00      7200      10000   
3  xmr       720     15        60       30   0.06   0.05         0      10000   
4  xmr       720     15        60       30   0.06   0.05      3600      10000   

   blocktime  ...  relativegain  adjustedrelativegain  gainstddev  \
0        120  ...     -0.822131             -0.823047         0.0   
1        120  ...     -0.829017             -0.829855         0.0   
2        120  ...     -0.830685             -0.831386         0.0   
3        120  ...     -0.772359             -0.773576         0.0   
4        120  ...     -0.775522             -0.776569         0.0   

   adjustedgainsteddev  secondsperblockstddev   finalheight   numreorgs  \
0                  0.0                    0.0  11000.100000  549.933333   
1                  0.0                    0.0  11000.000000  550.233333   
2                  0.0                    0.0  11000.100000  542.500000   
3                  0.0                    0.0  11000.133333  543.166667   
4                  0.0                    0.0  11000.000000  548.633333   

   smwinreorgs  didbetternaive  didbettertimeadjust  
0     0.059961             0.0                  0.0  
1     0.057940             0.0                  0.0  
2     0.056485             0.0                  0.0  
3     0.060811             0.0                  0.0  
4     0.062884             0.0                  0.0  

[5 rows x 23 columns]

In [ ]:
print(dataset['ltc'].shape)
dataset['ltc'].head()

(855, 22)


daa  period  offbyone  numsims  alpha  gamma  timewarp  numblocks  \
0  ltc    2016     False       30    0.2   0.00         0      10000   
1  ltc    2016     False       30    0.2   0.00      3600      10000   
2  ltc    2016     False       30    0.2   0.00      7200      10000   
3  ltc    2016     False       30    0.2   0.05         0      10000   
4  ltc    2016     False       30    0.2   0.05      3600      10000   

   blocktime  winratio  ...  relativegain  adjustedrelativegain  gainstddev  \
0        150  0.094300  ...     -0.528498             -0.546963           0   
1        150  0.092945  ...     -0.535273             -0.553404           0   
2        150  0.093595  ...     -0.532023             -0.549704           0   
3        150  0.101256  ...     -0.493719             -0.514090           0   
4        150  0.100324  ...     -0.498379             -0.518114           0   

   adjustedgainsteddev  secondsperblockstddev   finalheight    numreorgs  \
0                    0                      0  11000.366667  1522.933333   
1                    0                      0  11000.200000  1515.666667   
2                    0                      0  11000.200000  1518.433333   
3                    0                      0  11000.033333  1519.600000   
4                    0                      0  11000.233333  1522.266667   

   smwinreorgs  didbetternaive  didbettertimeadjust  
0     0.199167             0.0                  0.0  
1     0.200457             0.0                  0.0  
2     0.199436             0.0                  0.0  
3     0.200492             0.0                  0.0  
4     0.198415             0.0                  0.0  

[5 rows x 22 columns]

In [ ]:
temp=list(dataset.values())
final=pd.concat(temp,ignore_index=True)
final

daa  period offbyone  numsims  alpha  gamma  timewarp  numblocks  \
0      btc  2016.0     True       30   0.20   0.00         0      10000   
1      btc  2016.0     True       30   0.20   0.00      3600      10000   
2      btc  2016.0     True       30   0.20   0.00      7200      10000   
3      btc  2016.0     True       30   0.20   0.05         0      10000   
4      btc  2016.0     True       30   0.20   0.05      3600      10000   
...    ...     ...      ...      ...    ...    ...       ...        ...   
85270  ltc  2016.0    False       30   0.48   0.85      3600      10000   
85271  ltc  2016.0    False       30   0.48   0.85      7200      10000   
85272  ltc  2016.0    False       30   0.48   0.90         0      10000   
85273  ltc  2016.0    False       30   0.48   0.90      3600      10000   
85274  ltc  2016.0    False       30   0.48   0.90      7200      10000   

       blocktime  winratio  ...  npastblocks  lookback  mediantimepast  \
0            600  0.093275  ...          NaN       NaN             NaN   
1            600  0.092009  ...          NaN       NaN             NaN   
2            600  0.093352  ...          NaN       NaN             NaN   
3            600  0.101965  ...          NaN       NaN             NaN   
4            600  0.101375  ...          NaN       NaN             NaN   
...          ...       ...  ...          ...       ...             ...   
85270        150  0.867711  ...          NaN       NaN             NaN   
85271        150  0.861217  ...          NaN       NaN             NaN   
85272        150  0.870539  ...          NaN       NaN             NaN   
85273        150  0.877031  ...          NaN       NaN             NaN   
85274        150  0.860088  ...          NaN       NaN             NaN   

       navginterval  nmediantimespan  nmaxadjustup  nmaxadjustdown  \
0               NaN              NaN           NaN             NaN   
1               NaN              NaN           NaN             NaN   
2               NaN              NaN           NaN             NaN   
3               NaN              NaN           NaN             NaN   
4               NaN              NaN           NaN             NaN   
...             ...              ...           ...             ...   
85270           NaN              NaN           NaN             NaN   
85271           NaN              NaN           NaN             NaN   
85272           NaN              NaN           NaN             NaN   
85273           NaN              NaN           NaN             NaN   
85274           NaN              NaN           NaN             NaN   

       NPOWDampeningFactor  delay  outliers  
0                      NaN    NaN       NaN  
1                      NaN    NaN       NaN  
2                      NaN    NaN       NaN  
3                      NaN    NaN       NaN  
4                      NaN    NaN       NaN  
...                    ...    ...       ...  
85270                  NaN    NaN       NaN  
85271                  NaN    NaN       NaN  
85272                  NaN    NaN       NaN  
85273                  NaN    NaN       NaN  
85274                  NaN    NaN       NaN  

[85275 rows x 32 columns]

In [ ]:
order=['daa']
difficulty_par=list(original_dataset['difficulty_parameters'].apply(pd.Series).columns)
order.extend(difficulty_par)
results=list((original_dataset['results'].iloc[0])[0].keys())
order.extend(results)
order

['daa',
 'period',
 'offbyone',
 'npastblocks',
 'lookback',
 'mediantimepast',
 'navginterval',
 'nmediantimespan',
 'nmaxadjustup',
 'nmaxadjustdown',
 'NPOWDampeningFactor',
 'delay',
 'outliers',
 'numsims',
 'alpha',
 'gamma',
 'timewarp',
 'numblocks',
 'blocktime',
 'winratio',
 'adjustedwinning',
 'selfishsecondsperblock',
 'relativegain',
 'adjustedrelativegain',
 'gainstddev',
 'adjustedgainsteddev',
 'secondsperblockstddev',
 'finalheight',
 'numreorgs',
 'smwinreorgs',
 'didbetternaive',
 'didbettertimeadjust']

In [ ]:
final=final[order]
final.head()

daa  period offbyone  npastblocks  lookback  mediantimepast  navginterval  \
0  btc  2016.0     True          NaN       NaN             NaN           NaN   
1  btc  2016.0     True          NaN       NaN             NaN           NaN   
2  btc  2016.0     True          NaN       NaN             NaN           NaN   
3  btc  2016.0     True          NaN       NaN             NaN           NaN   
4  btc  2016.0     True          NaN       NaN             NaN           NaN   

   nmediantimespan  nmaxadjustup  nmaxadjustdown  ...  relativegain  \
0              NaN           NaN             NaN  ...     -0.533623   
1              NaN           NaN             NaN  ...     -0.539955   
2              NaN           NaN             NaN  ...     -0.533241   
3              NaN           NaN             NaN  ...     -0.490173   
4              NaN           NaN             NaN  ...     -0.493124   

   adjustedrelativegain  gainstddev  adjustedgainsteddev  \
0             -0.551429         0.0                  0.0   
1             -0.558573         0.0                  0.0   
2             -0.552124         0.0                  0.0   
3             -0.510206         0.0                  0.0   
4             -0.513377         0.0                  0.0   

   secondsperblockstddev   finalheight    numreorgs  smwinreorgs  \
0                    0.0  11000.200000  1531.433333     0.201882   
1                    0.0  11000.566667  1515.866667     0.198572   
2                    0.0  11000.233333  1514.266667     0.196568   
3                    0.0  11000.100000  1518.933333     0.201504   
4                    0.0  11000.133333  1517.300000     0.199095   

   didbetternaive  didbettertimeadjust  
0             0.0                  0.0  
1             0.0                  0.0  
2             0.0                  0.0  
3             0.0                  0.0  
4             0.0                  0.0  

[5 rows x 32 columns]